src/java/org/apache/cassandra/index/Index.java

```diff
@@ -364,6 +382,23 @@ default SSTableFlushObserver getFlushObserver(Descriptor descriptor, OperationTy
      */
     public AbstractType<?> customExpressionValueType();
 
+    /**
+     * If the index supports custom search expressions using the
+     * {@code SELECT * FROM table WHERE expr(index_name, expression)} syntax, this method should return a new
+     * {@link RowFilter.CustomExpression} for the specified expression value. Index implementations may provide their
+     * own implementations using method {@link RowFilter.CustomExpression#isSatisfiedBy(TableMetadata, DecoratedKey, Row)}
+     * to filter reconciled rows in the coordinator. Otherwise, the default implementation will accept all rows.
+     * See DB-2185 and DSP-16537 for further details.
+     *
+     * @param metadata the indexed table metadata
+     * @param value the custom expression value
+     * @return a custom index expression for the specified value
+     */
+    default RowFilter.CustomExpression customExpressionFor(TableMetadata metadata, ByteBuffer value)
```

Comment:
I think this method shouldn't be part of this ticket. IIRC we added it to allow indexes using custom expression to benefit from replica filtering protection (CASSANDRA-8272). However I think we are missing the part where `RowFilter.CustomExpression` uses it to delegate its instantiation to the index, the changes in `DataResolver` and the tests. IMO those changes should be done in a separate ticket, wdyt?

Reply:
removed